### PR TITLE
M0127CC-570 Add tutorial How to create your own MicroEJ platform on a custom board

### PR DESCRIPTION
**Pull request by gjadi-mej**
Imported from https://forum.microej.com/t/platforming-how-to-create-your-own-microej-platform-on-a-custom-board/48

### CODE REVIEW

**Comment by stephanemainchain** [comment]
Please replace RI with demo:
* A MicroEJ platform Reference Implementation (RI) => A MicroEJ demo platform
* platform RI => demo platform
*from the RI => from the demo

+ add a link at the end for platform validation https://github.com/MicroEJ/PlatformQualificationTools

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 1]
```diff
@@ -0,0 +1,130 @@
+How to create your own MicroEJ platform on a custom board
+=========================================================
+
+Introduction
+------------
+
+A MicroEJ architecture is a software package that includes the MicroEJ Runtime port to a specific target Instruction Set Architecture (ISA) and C compiler.
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:7]: global pass to use the glossary casing: MicroEJ *A*rchitecture - MicroEJ *P*latform - *S*imulator

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+It contains a set of libraries, tools and C header files. The MicroEJ architectures are provided by MicroEJ SDK.
+
+A MicroEJ platform reference implementation (RI) is a MicroEJ port for a custom board.
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:10]: Remove the term RI / MicroEJ platform reference implementation everywhere 
=> MicroEJ Platform

Here, better write something like: A MicroEJ Platform is a MicroEJ Architecture port for a custom device.

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+It contains the MicroEJ configuration and the BSP (C source files).
+The MicroEJ platform RI can be found on https://developer.microej.com.
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:12]: MicroEJ Corp. provides example of MicroEJ Platforms for evaluation boards on https://repository.microej.com/

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+We recommend reading the :ref:`section.firmware` section to get an overview of the firmware build flow.
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:14]: MicroEJ Firmware

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+The following document assumes the reader is familiar with the :ref:`platform-developer-guide`.
+
+Each platform reference implementation is specific to:
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:18]: Each MicroEJ Platform is specific to:

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+Import this platform in MicroEJ SDK.
+
+As the MCU, RTOS and compiler are the same, only the board specific code needs to be changed (external RAM, external oscillator, communication interfaces).
+
+Platform
+::::::::
+
+Required actions:
+
+* modify the .platform from the RI (``xxx-configuration`` project) in MicroEJ SDK to match your board features and its associated configuration (e.g. ``UI->Display``).
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:57]: In MicroEJ SDK
* modify the ``.platform`` from ``xxx-configuration`` project  to match your board features and its associated configuration (e.g. ``UI->Display``). <-- use guil label for menu items

File: Tutorials/How to create your own MicroEJ platform.rst [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  More details on which module does what may be gathered from the :ref:`platform-developer-guide`.
```
Comment (frivieremicroej) [inline Tutorials/How to create your own MicroEJ platform.rst:58]: does what may be gathered ???

File: Tutorials/index.rst [hunk 8]
```diff
@@ -7,6 +7,7 @@ Tutorials
    :maxdepth: 1
 
    applicationTrace
```
Comment (frivieremicroej) [inline Tutorials/index.rst:9]: @stephanemainchain decide which tuto between trace and this one will be included at first

**Comment by gjadi-mej** [comment]
@stephanemainchain @frivieremicroej I addressed your comments. Please see the latest changes.

File: Tutorials/How to create a MicroEJ Platform for a custom board.rst [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+* in the C IDE: start from an empty project that match with the MCU.
+
+Or to start from scratch a new MicroEJ Platform
+
+* In MicroEJ SDK: create the platform and edit (refer to the MicroEJ Platform as a model for implementation)
+* in the C IDE: start from an empty project and implement the drivers of each of the LL drivers API. 
+  
+  Make sure to link with:
+
+  * the ``microejruntime.lib`` that runs the JVM for the MCU Architecture
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom board.rst:110]: microejruntime.a (.lib is only for Keil specific compiler)

File: Tutorials/How to create a MicroEJ Platform for a custom board.rst [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+Platform Validation
+-------------------
+
+Use the `Platform Qualification Tools <https://github.com/MicroEJ/PlatformQualificationTools>`_ to qualify the MicroEJ Platform built.
+
+Further assistance needed
+-------------------------
+
+Please note that porting MicroEJ to a new board is also something that is part of our engineering services. Consider contacting sales@microej.com to request a quote.
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom board.rst:142]: board -> device (replace everywhere)

File: Tutorials/How to create a MicroEJ Platform for a custom board.rst [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+* same C compiler
+
+At this point, consider either to modify the closest MicroEJ Platform
+
+* In MicroEJ SDK: modify the platform configuration.
+* in the C IDE: start from an empty project that match with the MCU.
+
+Or to start from scratch a new MicroEJ Platform
+
+* In MicroEJ SDK: create the platform and edit (refer to the MicroEJ Platform as a model for implementation)
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom board.rst:105]: create the platform -> hyperlink to https://docs.microej.com/en/latest/PlatformDeveloperGuide/gettingstarted.html#create-a-new-microej-platform-configuration but with a manual anchor name: "_platform_creation".

File: Tutorials/How to create a MicroEJ Platform for a custom board.rst [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+These drivers are implemented in the platform BSP project. This project is edited in the C compiler IDE/dev environment (e.g. KEIL, GCC, IAR).
+E.g. the MicroUI library LED feature will require a ``LLLEDS.c`` that implements the native on/off IO drive.
+
+The following sections explain how to create a MicroEJ Platform for a custom board starting from an existing MicroEJ Platform.
+In the following, we assume that the new board hardware is validated and at least a trace output is available.
+It is also a good idea to run basic hardware test like:
+
+* Internal and external flash programming and verification
+* RAM 8/16/32 -bit read/write operations (internal and external if any)
+* EEMBC Coremark benchmark to verify the CPU/buses/memory/compiler configuration
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom board.rst:40]: add a ref to PQT Github for C validation (coremark, ...)

File: Tutorials/How to create a MicroEJ Platform for a custom board.rst [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+Several use cases are presented:
+
+A MicroEJ Platform is already available for the same MCU/RTOS/C Compiler
+------------------------------------------------------------------------
+
+This is the fastest way: the MicroEJ Platform is usually provided for a silicon vendor evaluation board.
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom board.rst:47]: keep board here
evaluation board - let as is
custom board => custom device

File: Tutorials/How to create a MicroEJ Platform for a custom board.rst [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+Import this platform in MicroEJ SDK.
+
+As the MCU, RTOS and compiler are the same, only the board specific code needs to be changed (external RAM, external oscillator, communication interfaces).
+
+Platform
+::::::::
+
+In MicroEJ SDK
+
+* modify the .platform from the MicroEJ Platform (``xxx-configuration`` project) to match the board features and its associated configuration (e.g. ``UI->Display``).
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom board.rst:57]: ``.platform``

**Comment by gjadi-mej** [comment]
@frivieremicroej Thanks for the review. See the latest commits.

File: Tutorials/How to create a MicroEJ Platform for a custom device.rst [hunk 15]
```diff
@@ -0,0 +1,143 @@
+How to create a MicroEJ Platform for a custom device
+====================================================
+
+Introduction
+------------
+
+A MicroEJ Architecture is a software package that includes the MicroEJ Runtime port to a specific target Instruction Set Architecture (ISA) and C compiler.
+It contains a set of libraries, tools and C header files. The MicroEJ Architectures are provided by MicroEJ SDK.
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom device.rst:8]: @stephanemainchain please review this part:

The MicroEJ Architectures are provided by MicroEJ SDK. 
=> 
MicroEJ Architectures for evaluation are available at https://repository.microej.com/architectures/. 

To get the MicroEJ Architecture for production counterpart, please contact MicroEJ support (mailto:support@microej.com) (you must own an active [Production License](https://docs.microej.com/en/latest/ApplicationDeveloperGuide/licenses.html#production-licenses)

File: Tutorials/How to create a MicroEJ Platform for a custom device.rst [hunk 16]
```diff
@@ -0,0 +1,143 @@
+How to create a MicroEJ Platform for a custom device
```
Comment (frivieremicroej) [inline Tutorials/How to create a MicroEJ Platform for a custom device.rst:1]: Please define a naming convention for tutorial rst files: Suggestion: `tutorial*.rst`
e.g.: `tutorialCustomPlatform.rst`

File: Tutorials/How to create a MicroEJ Platform for a custom device.rst (hunk 15 above)
Comment (gjadi-mej) [reply Tutorials/How to create a MicroEJ Platform for a custom device.rst:8]: @frivieremicroej  I find https://repository.microej.com/index.php with the "Firmware" filter easier to browse than https://repository.microej.com/architectures/.

File: Tutorials/How to create a MicroEJ Platform for a custom device.rst (hunk 16 above)
Comment (gjadi-mej) [reply Tutorials/How to create a MicroEJ Platform for a custom device.rst]: Done. Named `tutorialCreateCustomPlatform.rst`.

File: Tutorials/tutorialCreateCustomPlatform.rst [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+Platform
+::::::::
+
+In MicroEJ SDK
+
+* modify the ``.platform`` from the MicroEJ Platform (``xxx-configuration`` project) to match the device features and its associated configuration (e.g. ``UI->Display``).
+
+  .. image:: images/tuto_microej_platform_how_to_select_modules.PNG
+
+  More details available modules can be found in the :ref:`platform-developer-guide`.
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateCustomPlatform.rst:62]: more details *on* available modules

File: Tutorials/How to create a MicroEJ Platform for a custom device.rst (hunk 15 above)
Comment (frivieremicroej) [reply Tutorials/How to create a MicroEJ Platform for a custom device.rst]: No use this link,  https://repository.microej.com/architectures/ with the changelog + I just populated evaluation architectures 1 week ago

File: Tutorials/index.rst (hunk 8 above)
Comment (frivieremicroej) [reply Tutorials/index.rst:9]: (PR https://github.com/MicroEJ/docs/pull/34 must be merged before)

Comment (frivieremicroej) [reply Tutorials/index.rst:9]: image

File: Tutorials/tutorialCreateCustomPlatform.rst [hunk 18]
```diff
@@ -0,0 +1,143 @@
+How to create a MicroEJ Platform for a custom device
```
Comment (frivieremicroej) [inline Tutorials/tutorialCreateCustomPlatform.rst:1]: @stephanemainchain I suggest to remove "How To" to reduce the title

image

Comment (frivieremicroej) [reply Tutorials/tutorialCreateCustomPlatform.rst:1]: Strange End of sentence jumps to a title 

image

Comment (gjadi-mej) [reply Tutorials/tutorialCreateCustomPlatform.rst:1]: @frivieremicroej Should we remove the "How To" from the title in the index page only or for both the title in the index page AND the title on the actual tutorial's page?

Comment (frivieremicroej) [reply Tutorials/tutorialCreateCustomPlatform.rst:1]: I suggest to simply remove "How to" for both

Comment (gjadi-mej) [reply Tutorials/tutorialCreateCustomPlatform.rst]: @frivieremicroej Done.

Comment (stephanemainchain) [reply Tutorials/tutorialCreateCustomPlatform.rst]: Agreed.

**Pull request merged**